File: implementations/fastapi/examples/src/order/application/command/cancel_order_handler.py
from dataclasses import dataclass

from ...domain.errors import OrderNotFoundError
from ...domain.repository import OrderRepository


@dataclass
class CancelOrderCommand:
    order_id: str
    reason: str


class CancelOrderHandler:

    def __init__(self, repo: OrderRepository) -> None:
        self._repo = repo

    async def execute(self, cmd: CancelOrderCommand) -> None:
        order = await self._repo.find_by_id(cmd.order_id)
        if order is None:
            raise OrderNotFoundError(cmd.order_id)
        order.cancel(cmd.reason)
        await self._repo.save(order)

File: implementations/go/examples/internal/application/command/cancel_order_handler.go
package command

import (
	"context"
	"fmt"

	"github.com/example/order-service/internal/domain/order"
)

type CancelOrderCommand struct {
	OrderID string
	Reason  string
}

type CancelOrderHandler struct {
	repo order.Repository
}

func NewCancelOrderHandler(repo order.Repository) *CancelOrderHandler {
	return &CancelOrderHandler{repo: repo}
}

func (h *CancelOrderHandler) Handle(ctx context.Context, cmd CancelOrderCommand) error {
	o, err := h.repo.FindByID(ctx, cmd.OrderID)
	if err != nil {
		return fmt.Errorf("cancel order: %w", err)
	}
	if err := o.Cancel(cmd.Reason); err != nil {
		return err
	}
	return h.repo.Save(ctx, o)
}
